images/cluster-setup/install.sh

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env bash
+
+# Copyright 2022 The pipelines-service Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+usage() {
+
+    printf "Usage: DATA_DIR="/workspace" ./register.sh \n\n"
+
+    # Parameters
+    printf "The following parameters need to be passed to the script: \n"
+    printf "DATA_DIR: the location of the cluster files \n"
+}
+
+prechecks () {
+    DATA_DIR=${DATA_DIR:-}
+    if [[ -z "${DATA_DIR}" ]]; then
+        printf "DATA_DIR not set\n\n"
+        usage
+	      exit 1
+    fi
+}
+
+# populate clusters with the cluster names taken from the kubeconfig
+# populate contexts with the context name taken from the kubeconfig
+# populate kubeconfigs with the associated kubeconfig for each cluster name
+# only consider the first context for a specific cluster
+get_clusters() {
```

Comment:
That seems a bit complex. Couldn't it be simplified by retrieving the list of clusters like [this](https://github.com/Roming22/pipelines-service/blob/feat/register-gitops/images/argocd/register.sh#L78),

Reply:
you can have multiple context and multiple clusters in a kubeconfig file

Reply:
I'm having difficulty following your code in the link you shared.
- In the process_clusters() func, why are we checking for kustomization.yaml by name?
- In the same func, we are passing cluster_dir to process_cluster() func. Why are we passing this path?
  cluster_dir returns something like this - /home/bnr/workspace/pipelines-service/gitops/environment/compute/base/argocd-rbac
- And then in process_cluster() func, you are checking for ${cluster_name}.yaml which may not be the case. Users would usually have *.kubeconfig files under credentials.
- Also, I am not sure if it is handling multiple cluster information inside one kubeconfig file.

Reply:
@fgiloux I don't believe that having multiple clusters/contexts per kubeconfig would help managing the repository.  If a process can edit a kubeconfig to add/remove a context or cluster, we run into the risk of having a conflict on that file. I think that it would be best to have a single file per cluster/context.

@bnallapeta Every cluster has it's own kubeconfig, under `credential/compute/$cluster_name.yaml`. Every cluster has it's gitops configuration under `environment/compute/.../$cluster_name/kustomization.yaml`. Therefore:

1. To list the clusters that need to be setup, we just need to have the list of directories in `environment/compute` that have a `kustomization.yaml`.
2. $cluster_dir has 2 informations. The name of the cluster we want to setup, and the path holding the `kustomization.yaml`. We need the first info to get the right kubeconfig from `credentials/compute`, and the second to do the `kubectl apply`.
3. I used the yaml extension for the kubeconfig, because that's what they are, That might be a mistake, so let's put that aside. We would however require that for any `environment/compute/.../$cluster_name/kustomization.yaml` there is a corresponding `credential/compute/$cluster_name.[yaml|kubeconfig]`.
4. At the moment I don't see being able to handle multiple clusters/contexts in a single kubeconfig as desirable. As per my comment to Frederic, I think it could open the door to conflict issues as PR opened by humans and robots could start modifying the same area of the same file. With a single file per cluster/context, the most likely scenario is the file will either be added or deleted, and most likely will never be modified.

Reply:
@Roming22 we are not creating the kubeconfig files. They are provided by the persons or tools making0 the workload clusters available. When you log into multiple clusters or with multiple users without changing the KUBECONFIG environment variable you end up with multiple clusters and contexts in the kubeconfig file. Sure we could put additional restrictions to what is accepted as kubeconfig but I am missing the urge for it as Bharath already has code that can deal with it.